### PR TITLE
Fixes #54550

### LINKED ISSUE #38571: The fact variable {{ ansible_nodename}} has an extra dot at the end on the windows which without domain name
##### ISSUE TYPE

 - Bug Report

##### COMPONENT NAME

module

##### ANSIBLE VERSION

```
ansible 2.4.2.0
```

##### OS / ENVIRONMENT

Windows

The infomation from GetIPGlobalProperties() as below:
```
HostName      : jobwin8
DomainName    :
NodeType      : Hybrid
DhcpScopeName :
IsWinsProxy   : False
```
##### SUMMARY

When I try to use the fact {{ ansible_nodename}} of the Windows machine which without DomainName, I found that {{ ansible_nodename}} is incorrect. It has an extra dot at the end. For example, my hostname is "jobwin8" without domain name, but the {{ ansible_nodename}} and {{ ansible_fqdn}} is "jobwin8.". I think it is not right. On the Redhat OS the fact variable is right.
I found that in the code ansible/lib/ansible/modules/windows/setup.ps1, the {{ ansible_nodename}} and {{ ansible_fqdn} is be joined by hostname , dot and domain name. When the DomainName is empty, the dot also be added.
```
    $ansible_facts += @{
        ansible_architecture = $win32_os.OSArchitecture
        ansible_domain = $ip_props.DomainName
        ansible_fqdn = ($ip_props.Hostname + "." + $ip_props.DomainName)
        ansible_hostname = $env:COMPUTERNAME
        ansible_kernel = $osversion.Version.ToString()
        ansible_nodename = ($ip_props.HostName + "." + $ip_props.DomainName)
```
##### STEPS TO REPRODUCE

Output the hostvars for the windows machine which without domain name.

```yaml
---
  - debug:
      var: hostvars
```

##### ACTUAL RESULTS

```

```

### LINKED ISSUE #54550: Ansible Windows Setup module should not use NETBIOS name for ansible_hostname
##### SUMMARY

There is a bug in the Windows Setup module of Ansible. Ansible uses the `$env:COMPUTERNAME` to determine `ansible_hostname`, see here:

https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/windows/setup.ps1#L323

According to https://stackoverflow.com/a/1233458/474034 the computer name resolves to the local NETBIOS name and can only be 15 characters long:

> [The NETBIOS] name is restricted to 15 characters and only visible on the LAN.

This means, that when I have a hostname on AWS EC2 that has the form ip-172-34-55-231, which is 16 characters long, then the `ansible_hostname` variable will have the last value truncated, e.g. (`ip-172-34-55-23`).

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
windows/setup

##### ANSIBLE VERSION
latest

##### OS / ENVIRONMENT
Windows

##### STEPS TO REPRODUCE
1. Create a playbook an invoke the setup module.
2. Prepare a windows machine which has a hostname that is longer than 15 characters.
3. Observer that on this windows machine the COMPUTERNAME is truncated to 15 characters but the hostname has the correct value.

       PS C:\Users\admin> [System.Net.Dns]::GetHostName()
       ip-172-34-55-231
       PS C:\Users\admin> hostname
       ip-172-34-55-231
       PS C:\Users\admin> echo $env:COMPUTERNAME
       IP-172-34-55-23

4. Run the playbook on a windows machine and observe that the `ansible_hostname` uses the COMPUTERNAME which is truncated to 15 characters.

##### EXPECTED RESULTS
The ansible_hostname should have the correct hostname which is not truncated. Proposed solution: use `[System.Net.Dns]::GetHostName()` to determine the hostname:

    ansible_hostname = [System.Net.Dns]::GetHostName()

##### ACTUAL RESULTS
The hostname in `ansible_hostname` is truncated because `$env:COMPUTERNAME` is used by the setup module.

### PR DESCRIPTION
**Pull request by ShachafGoldstein**
##### SUMMARY
Fixes #54550 
Fixes: #38571

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
setup.ps1

### CODE REVIEW

**Comment by jhawkesworth** [comment]
Not tested but this seems like a good change to have to me.
Would be nice to perhaps change the win_hostname module and tests for win_hostname (test/integration/targets/win_hostname/tasks/ ) module to exercise this and ensure non-netbios hostnames are managed there as well (probably best as a separate PR).

**Comment by ShachafGoldstein** [comment]
> Not tested but this seems like a good change to have to me.
> Would be nice to perhaps change the win_hostname module and tests for win_hostname (test/integration/targets/win_hostname/tasks/ ) module to exercise this and ensure non-netbios hostnames are managed there as well (probably best as a separate PR).

There is a separate issue regarding win_hostname and netbios names, ill tackle that one soon and ill add the mentioned tests there (https://github.com/ansible/ansible/issues/55283)

**Comment by ShachafGoldstein** [comment]
Change to use Win32_ComputerSystem and add ansible_netbios_name, check for name dependency like fqdn facts

File: lib/ansible/modules/windows/setup.ps1 [hunk 1]
```diff
@@ -319,10 +319,11 @@ if($gather_subset.Contains('platform')) {
     $ansible_facts += @{
         ansible_architecture = $win32_os.OSArchitecture
         ansible_domain = $ip_props.DomainName
```
Comment (dagwieers) [inline lib/ansible/modules/windows/setup.ps1]: $ip_props is still used here.

**Review by dagwieers — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/windows/setup.ps1 (hunk 1 above)
Comment (ShachafGoldstein) [reply lib/ansible/modules/windows/setup.ps1]: Fixed

File: lib/ansible/modules/windows/setup.ps1 [hunk 2]
```diff
@@ -307,7 +307,7 @@ if($gather_subset.Contains('memory')) {
 if($gather_subset.Contains('platform')) {
     $win32_cs = Get-LazyCimInstance Win32_ComputerSystem
     $win32_os = Get-LazyCimInstance Win32_OperatingSystem
-    $ip_props = [System.Net.NetworkInformation.IPGlobalProperties]::GetIPGlobalProperties()
+    #$ip_props = [System.Net.NetworkInformation.IPGlobalProperties]::GetIPGlobalProperties()
```
Comment (jborean93) [inline lib/ansible/modules/windows/setup.ps1]: No use keeping this commented, we have git history if we need to see what was originally there.

File: lib/ansible/modules/windows/setup.ps1 [hunk 3]
```diff
@@ -318,11 +318,12 @@ if($gather_subset.Contains('platform')) {
 
     $ansible_facts += @{
         ansible_architecture = $win32_os.OSArchitecture
-        ansible_domain = $ip_props.DomainName
-        ansible_fqdn = ($ip_props.Hostname + "." + $ip_props.DomainName)
-        ansible_hostname = $env:COMPUTERNAME
+        ansible_domain = $win32_cs.Domain.Substring($win32_cs.Workgroup.length)
```
Comment (jborean93) [inline lib/ansible/modules/windows/setup.ps1]: What's the purpose of the Substring here?

File: lib/ansible/modules/windows/setup.ps1 [hunk 4]
```diff
@@ -318,11 +318,12 @@ if($gather_subset.Contains('platform')) {
...
+        ansible_fqdn = ($win32_cs.DNSHostname + "." + $win32_cs.Domain.Substring($win32_cs.Workgroup.length))
```
Comment (jborean93) [inline lib/ansible/modules/windows/setup.ps1]: This still has the same issue as before where a non-domain host would be `HOSTNAME.` (with the ending .). We should conditionally add the domain part with `.` to the FQDN if it's not an empty string. Also the same thing here about the Substring for the domain name.

File: lib/ansible/modules/windows/setup.ps1 [hunk 5]
```diff
@@ -318,11 +318,12 @@ if($gather_subset.Contains('platform')) {
...
+        ansible_hostname = $win32_cs.DNSHostname
+        ansible_netbios_name = $win32_cs.Name
         ansible_kernel = $osversion.Version.ToString()
-        ansible_nodename = ($ip_props.HostName + "." + $ip_props.DomainName)
+        ansible_nodename = ($win32_cs.DNSHostname + "." + $win32_cs.Domain.Substring($win32_cs.Workgroup.length))
```
Comment (jborean93) [inline lib/ansible/modules/windows/setup.ps1]: If we are reusing values like before it could be beneficial to store that in a var and just set the var here instead of calculating it twice.

File: changelogs/fragments/54550-Change-windows-ansible-hostname-setup.yml [hunk 6]
```diff
@@ -0,0 +1,2 @@
+bugfixes:
+  - "setup.ps1 - Change $env:COMPUTERNAME to [System.Net.Dns]::GetHostName(), to support non NETBIOS compliant hostnames  (https://github.com/ansible/ansible/issues/54550)"
```
Comment (jborean93) [inline changelogs/fragments/54550-Change-windows-ansible-hostname-setup.yml]: Needs to be updated, it might be better to leave our implementation changes and just say what was fixed, i.e. non NETBIOS comliant hostnames.

**Review by jborean93 — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/windows/setup.ps1 (hunk 3 above)
Comment (ShachafGoldstein) [reply lib/ansible/modules/windows/setup.ps1]: A trick to get it to give empty string in a work group and the domain name in domain
to make sure the change backward compatible

File: lib/ansible/modules/windows/setup.ps1 (hunk 4 above)
Comment (ShachafGoldstein) [reply lib/ansible/modules/windows/setup.ps1]: Kept the dot or backward compatibility, so you wish to remove it if not needed?

File: lib/ansible/modules/windows/setup.ps1 (hunk 5 above)
Comment (ShachafGoldstein) [reply lib/ansible/modules/windows/setup.ps1]: Changed

File: lib/ansible/modules/windows/setup.ps1 (hunk 4 above)
Comment (ShachafGoldstein) [reply lib/ansible/modules/windows/setup.ps1]: Last commit removes it

File: lib/ansible/modules/windows/setup.ps1 (hunk 3 above)
Comment (jborean93) [reply lib/ansible/modules/windows/setup.ps1]: In my testing `Domain` is not set for a workgroup host, this means that this does nothing for those scenarios. Are you seeing something else here that makes this trick necessary?

Comment (ShachafGoldstein) [reply lib/ansible/modules/windows/setup.ps1]: `Get-CimInstance win32_computersystem -Property domain,Workgroup | select domain,Workgroup`
On my laptop this gives the workgroup name in both, so if this were to run as is without the trick I would get a domain name, but there is no domain

File: lib/ansible/modules/windows/setup.ps1 [hunk 7]
```diff
@@ -307,7 +307,12 @@ if($gather_subset.Contains('memory')) {
 if($gather_subset.Contains('platform')) {
     $win32_cs = Get-LazyCimInstance Win32_ComputerSystem
     $win32_os = Get-LazyCimInstance Win32_OperatingSystem
-    $ip_props = [System.Net.NetworkInformation.IPGlobalProperties]::GetIPGlobalProperties()
+    $domainSuffix = $win32_cs.Domain.Substring($win32_cs.Workgroup.length)
+
+    if ($domainSuffix -ne "")
+    {s
```
Comment (jborean93) [inline lib/ansible/modules/windows/setup.ps1]: Extra s

File: lib/ansible/modules/windows/setup.ps1 [hunk 8]
```diff
@@ -307,7 +307,12 @@ if($gather_subset.Contains('memory')) {
 if($gather_subset.Contains('platform')) {
     $win32_cs = Get-LazyCimInstance Win32_ComputerSystem
     $win32_os = Get-LazyCimInstance Win32_OperatingSystem
-    $ip_props = [System.Net.NetworkInformation.IPGlobalProperties]::GetIPGlobalProperties()
+    $domainSuffix = $win32_cs.Domain.Substring($win32_cs.Workgroup.length)
```
Comment (jborean93) [inline lib/ansible/modules/windows/setup.ps1]: I would follow the existing variable standard which is `snake_case`, e.g. `$domain_suffix`.

File: lib/ansible/modules/windows/setup.ps1 [hunk 9]
```diff
@@ -318,11 +323,12 @@ if($gather_subset.Contains('platform')) {
 
     $ansible_facts += @{
         ansible_architecture = $win32_os.OSArchitecture
-        ansible_domain = $ip_props.DomainName
-        ansible_fqdn = ($ip_props.Hostname + "." + $ip_props.DomainName)
-        ansible_hostname = $env:COMPUTERNAME
+        ansible_domain = $domainSuffix
```
Comment (jborean93) [inline lib/ansible/modules/windows/setup.ps1]: You are using the value which may have `.` at the start, currently we just return the domain part and not the leading `.`.

File: lib/ansible/modules/windows/setup.ps1 [hunk 10]
```diff
@@ -318,11 +323,12 @@ if($gather_subset.Contains('platform')) {
...
+        ansible_fqdn = ($win32_cs.DNSHostname + "." + $domainSuffix)
```
Comment (jborean93) [inline lib/ansible/modules/windows/setup.ps1]: This will still have the same issue where if there is no domain suffix then the value is `"hostname."`. We need to fix it up so it only return the hostname if there is no domain. Also because `$domainSuffix` has the leading `.` we would now have `hostname..domain.com`.

**Review by jborean93 — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/windows/setup.ps1 [hunk 11]
```diff
@@ -307,7 +307,9 @@ if($gather_subset.Contains('memory')) {
 if($gather_subset.Contains('platform')) {
     $win32_cs = Get-LazyCimInstance Win32_ComputerSystem
     $win32_os = Get-LazyCimInstance Win32_OperatingSystem
-    $ip_props = [System.Net.NetworkInformation.IPGlobalProperties]::GetIPGlobalProperties()
+    $domain_suffix = $win32_cs.Domain.Substring($win32_cs.Workgroup.length)
+
+    $domain_suffix = ".$domain_suffix"
```
Comment (jborean93) [inline lib/ansible/modules/windows/setup.ps1]: This will still be an issue when `$domain_suffix` is an empty string.

File: .gitignore [hunk 12]
```diff
@@ -19,6 +19,7 @@ rpm-build
 *.iml
 #VSCode stuff..
 .vscode/
+.favorites.json
```
Comment (jborean93) [inline .gitignore]: Should be added in a separate PR

**Review by jborean93 — CHANGES_REQUESTED** [review]
Still some issues you are best of doing something like;

```
$domain_name = $win32_cs.Domain.Substring($win32_cs.Workgroup.length)
$dns_hostname = $win32_cs.DNSHostname
$fqdn = $dns_hostname

if ($domain_name -ne "") {
    $fqdn = "$($dns_hostname).$($domain_suffix)"
}
```

From there you can just set `ansible_domain`, `ansible_fqdn`, `ansible_hostname`, and `ansible_nodename` based on these values. The only one that is special is `ansible_fqdn` which has the domain added if it exists.

File: lib/ansible/modules/windows/setup.ps1 (hunk 11 above)
Comment (ShachafGoldstein) [reply lib/ansible/modules/windows/setup.ps1]: that is why I changed `ansible_domain = $domain_suffix.Substring(1)`

File: .gitignore (hunk 12 above)
Comment (ShachafGoldstein) [reply .gitignore]: bad configuration on my side, reverted

File: lib/ansible/modules/windows/setup.ps1 (hunk 11 above)
Comment (jborean93) [reply lib/ansible/modules/windows/setup.ps1]: Yea I realised that later but we still have the issue with the fqdn value as it will have the `.` as a suffix if no domain is set.

Comment (ShachafGoldstein) [reply lib/ansible/modules/windows/setup.ps1]: right, did some mix and match.
now it should be ok, empty will stay empty and domain will get its dot

working so late has its ups and downs

File: lib/ansible/modules/windows/setup.ps1 [hunk 13]
```diff
@@ -318,11 +323,12 @@ if($gather_subset.Contains('platform')) {
 
     $ansible_facts += @{
         ansible_architecture = $win32_os.OSArchitecture
-        ansible_domain = $ip_props.DomainName
-        ansible_fqdn = ($ip_props.Hostname + "." + $ip_props.DomainName)
-        ansible_hostname = $env:COMPUTERNAME
+        ansible_domain = $domain_suffix
```
Comment (jborean93) [inline lib/ansible/modules/windows/setup.ps1:327]: We still have the same issue here, the value will be `.domain.com` when it should be `domain.com`.

**Review by jborean93 — CHANGES_REQUESTED** [review]
The same issue with `ansible_domain` is now back

File: lib/ansible/modules/windows/setup.ps1 (hunk 13 above)
Comment (ShachafGoldstein) [reply lib/ansible/modules/windows/setup.ps1:327]: my brain is a pudding, sec

**Comment by ShachafGoldstein** [comment]
@jborean93 - take a look at this one, if you still see an issue ill come back to it in the morning

File: lib/ansible/modules/windows/setup.ps1 [hunk 14]
```diff
@@ -307,7 +307,12 @@ if($gather_subset.Contains('memory')) {
 if($gather_subset.Contains('platform')) {
     $win32_cs = Get-LazyCimInstance Win32_ComputerSystem
     $win32_os = Get-LazyCimInstance Win32_OperatingSystem
-    $ip_props = [System.Net.NetworkInformation.IPGlobalProperties]::GetIPGlobalProperties()
+    $domain_suffix = $win32_cs.Domain.Substring($win32_cs.Workgroup.length)
```
Comment (jborean93) [inline lib/ansible/modules/windows/setup.ps1:310]: Now `ansible_domain` will fail in a workgroup because `$domain_suffix` is an empty string and `.Substring(1)` will fail. What I suggest you do is;

```suggestion
    $domain = $win32_cs.Domain.Substring($win32_cs.Workgroup.length)
    $fqdn = $win32_cs.DNSHostname
    if ($domain -ne "") {
        $fqdn += ".$domain"
    }
    
    try {
        $ansible_reboot_pending = Get-PendingRebootStatus
    } catch {
        # fails for non-admin users, set to null in this case
        $ansible_reboot_pending = $null
    }

    $ansible_facts += @{
        ansible_architecture = $win32_os.OSArchitecture
        ansible_domain = $domain
        ansible_fqdn = $fqdn
        ansible_hostname = $win32_cs.DNSHostname
        ansible_netbios_name = $win32_cs.Name
        ansible_kernel = $osversion.Version.ToString()
        ansible_nodename = $fqdn
```

What this means is;

* `ansible_domain`: Is the value returned from `$win32_cs.Domain.Substring($win32_cs.Workgroup.length)` which in a domain environment is the domain part and a workgroup is an empty string
* `ansible_fqdn`: Is just the DNS hostname in a workgroup, if a domain is present it is the DNS hostname with `.$domain` appended
* `ansible_nodename`: Is the same as fqdn

This solves the issue where `ansible_domain` is an empty string for a workgroup and contains the domain part without the leading `.` for an actual domain environment. Also the fqdn will not contain the `.` suffix in a workgroup as we only add that if the domain value is not an empty string.

_Note; I'm not sure how well GitHub suggestions work for multiple lines, you may need to manually apply this, or modify it how you like._

**Review by jborean93 — CHANGES_REQUESTED** [review]

**Review by jborean93 — APPROVED** [review]
woot!

**Comment by ShachafGoldstein** [comment]
bot_status
